### ISSUE TITLE
There's still a need to bump the memcache size

### ISSUE THREAD

**Issue by jidanni** [openstreetmap/operations#1107]
Hello. In https://github.com/openstreetmap/openstreetmap-website/issues/2457 I was told to open an issue here. But as it is getting a little over my head, I will just leave this here.

**Comment by tomhughes** [comment]
There is no evidence at all in the graphs that this in fact an issue. I definitely see the issue that you are referring to but I am unable it as all evidence says it shouldn't be down to memcache.

**Comment by tbertels** [comment]
Could these sessions disconnections be caused by server restarts or does the server never restart?

**Comment by mmd-osm** [comment]
I don't see any server restart in the stats, at least for the last 6 months: https://prometheus.openstreetmap.org/d/l4zgNUdMz/memcached?orgId=1&refresh=1m&from=now-6M&to=now

Also, the OP didn't provide any details how frequently they have to log in again. There might be external factors, like cookies being removed by the browser or some browser extension, etc.

**Comment by jidanni** [comment]
I thought everybody else also has to login again at least once every three or four days.
Maybe it's because I use various browsers on various devices. But why on the same device do I need to login again after three or four days?
Anyways welcome to check the logs to see why user jidanni has to login again so often.

**Comment by tbertels** [comment]
Which stat do you use to check if the server restarted?
Aren't these sudden drops in memory usage symptoms of a server restart?
Note that the dates are in the format month/day.
![Copie d'écran_20240707_143049m](https://github.com/openstreetmap/operations/assets/3265870/5b15e0e0-46b1-4fbb-ad9c-6c5eff760d45)

**Comment by mmd-osm** [comment]
Ah, the link wasn't that helpful. There are about 11 memcached instances overall. However, for the 3 frontend servers, only 3 memcached instances (spike-06 ... spike-08) are relevant. Items in cache and memory usage are fairly stable for these three.

https://prometheus.openstreetmap.org/d/l4zgNUdMz/memcached?orgId=1&refresh=1m&from=now-6M&to=now&var-instance=spike-06&var-instance=spike-07&var-instance=spike-08

**Comment by tomhughes** [comment]
A restart would indeed lose all sessions but as @mmd-osm says it's only those three machines that we're talking about here and they last restarted in November last year:

![image](https://github.com/openstreetmap/operations/assets/147741/bec16a80-0ab6-4b15-b83c-cb58b5ef14bb)

At that time it took nearly two months for the caches to fill up which suggests that it should take about that long for things to get expired unless there has been a significant increase in the cache usage since.

**Comment by pnorman** [comment]
The eviction rate has increased since November but it hasn't consisntently bee more than double. commands/second has remaind the same

**Comment by tbertels** [comment]
I logged back in 5 days ago: 1 day later my session was still active but today I'm logged out.
We can also [see a dip today](https://prometheus.openstreetmap.org/d/l4zgNUdMz/memcached?orgId=1&refresh=1m&from=now-6M&to=now&var-instance=spike-06&var-instance=spike-07&var-instance=spike-08) from ~100 millions items in cache to ~66 millions.

I suggest to store the sessions in the DB and use memcache only to speed up sessions check for frequently used sessions.

**Comment by tomhughes** [comment]
One of the machines was rebooted yesterday while fighting the DDOS so 1/3 of the the cache entries were lost.

**Comment by mmd-osm** [comment]
I'm wondering how many of these entries originate from CGImap (key prefix would be "cgimap:"). For some reason, these entries have the expiration value set to 0 (unlimited). This doesn't make a whole lot of sense for rate limiting requests, where the exact timestamp would be known upfront when these entries become irrelevant.

**Comment by mmd-osm** [comment]
At least when testing locally, I've noticed that every anonymous user creates a rails session without expiry (that's the "0" in "1 0 73" below), whereas logged in users have an entry with 4-5 weeks expiration.

Anonymous user sessions:

```
/usr/share/memcached/scripts/memcached-tool localhost:11211 dump 
Dumping memcache contents
add rails:session:2::2d28d018bdda81f05bae57ba42ee200a7a14af6df74134bb93ee82f99bf7baab 1 0 73
{I"_csrf_token:EFI"096xa2ms9DVncEF7CBUeBJ0wP9VYJrKO6lzxqDomep74;F
```

Logged in user:

1723288155 = Sat Aug 10 13:09:15 CEST 2024

```
add rails:session:2::2d28d018bdda81f05bae57ba42ee200a7a14af6df74134bb93ee82f99bf7baab 1 1723288155 200
{	I"_csrf_token:EFI"096xa2ms9DVncEF7CBUeBJ0wP9VYJrKO6lzxqDomep74;FI"	user;FiI"fingerprint;FI"E....
```

**Comment by tomhughes** [comment]
Expiry shouldn't really matter that much because anything that isn't used just moves down the LRU list and gets discarded eventually when we need space for a new entry.

Logged in sessions (with "remember me" checked) do get an expiry of 28 days which matches the cookie expiry while other sessions (not logged in and logged in without "remember me" checked) actually don't have an expiry but issue a session cookie that expires when the browser is closed.

**Comment by mmd-osm** [comment]
First of all, I find it a bit difficult to reason about the logged in sessions based on Prometheus stats, in particular after how many days these entries would be discarded.

memcached has an [LRU crawler](https://memcached.org/blog/modern-lru/) which reclaims expired entries even before they're reaching the end of the LRU list. With a non zero TTL, we might get rid of many "non-logged in user" entries early on, before they might evict "logged in user" entries.

**Comment by mmd-osm** [comment]
At the current growth rate, we will likely see some evictions in about 10 days (=21 days after last memcached restart).

@jidanni : did you notice any issues with lost login sessions in the last 8-9 days? If so, it can’t be memcached related…

**Comment by tomhughes** [comment]
It's not that simple because only one machine was reset I think? So only keys which hash to that machine are currently exempt from being evicted.

**Comment by mmd-osm** [comment]
I think spike-06..08 were all restarted, the aggregated cache items count on Prometheus shows 0 entries about 10 days ago.

**Comment by jidanni** [comment]
@mmd-osm rather than using my misty memory, 
surely there must be some internal logs you can check regarding  (user: jidanni)
that can give you precise details.

**Comment by mmd-osm** [comment]
We want to hear from you first hand, as you’ve also raised the issue. Misty memory is ok. If you say it hasn’t bothered you recently then that’s good enough for now.

**Comment by jidanni** [comment]
Okay. I will remember next time to report each and every incident right here to the thread.

**Comment by jidanni** [comment]
Okay. Just had to log in again as you can see in your logs perhaps.

**Comment by mmd-osm** [comment]
Thank you for the feedback. This is not completely unexpected. Evicting entries started again on August 1st, even a bit sooner than estimated.

**Comment by jidanni** [comment]
On a laptop I hadn't used in five days:
Had to login again to OSM.
But didn't need to login again to GitHub to add this comment.

**Comment by mmd-osm** [comment]
At least 5 other users have reported the same issue in https://community.openstreetmap.org/t/osm-webseite-standiges-login-notig/120072

spike-0[6-8] are seeing some cache evictions since a few days again:

<details>

![image](https://github.com/user-attachments/assets/8441fe34-a8b6-4caa-9fd7-4d882f060dd8)

![image](https://github.com/user-attachments/assets/ed28eda4-448e-4e55-a9f7-bcb96af66eb9)

</details>

Following up on my previous comment to get rid of anonymous sessions as early as possible, we should check if adding an `expire_after` parameter in session_store.rb could help. 

```ruby
if Settings.key?(:memcache_servers)
  Rails.application.config.session_store :mem_cache_store, 
                 :memcache_server => Settings.memcache_servers,
                 :namespace => "rails:session", 
                 :key => "_osm_session", 
                 :same_site => :lax,
                 :expire_after => 3600
else
  Rails.application.config.session_store :cache_store, :key => "_osm_session", :cache => ActiveSupport::Cache::MemoryStore.new, :same_site => :lax
end
```

I spent some time debugging the Dalli client, and it seems that this parameter is used for anonymous sessions, while named sessions still expire after 30 days.

 ~/.rbenv/versions/3.3.2/lib/ruby/gems/3.3.0/gems/dalli-3.2.8/lib/dalli/client.rb is where the various getter and setter methods are, and where the TTL can be checked for each use case. 

I've found some other comments, where the expire_after parameter was explicitly set via rack session option. It seems the 30 days are inherited from the cookie duration. So this extra step didn't seem to be required for named sessions.

```ruby
    env['rack.session.options'].merge! expire_after: 2592000
```

**Comment by mmd-osm** [comment]
I've adjusted the Gitlab code a bit to work with the osm website: https://github.com/mmd-osm/openstreetmap-website/commit/6fd17a0426d180836c42961c67d5146a27005456

It's more of a proof of concept at this time, to demo the idea. I can create a PR to continue the discussion, if needed.

**Comment by jidanni** [comment]
Today had to login again.

**Comment by jidanni** [comment]
Today had to login again too.

**Comment by jidanni** [comment]
Also today Wed Jan 29 03:07:20 AM UTC 2025
had to login again too. Yes, "Remember me" checked last time as always I do.

**Comment by tbertels** [comment]
I had to login again on 2025-02-26 and today (2025-02-29).

**Comment by jidanni** [comment]
> I had to login again on 2025-02-26 and today (2025-02-29).

Me too, today 2025/2/3.

Hey wait, your clock is a month ahead.

Anyway,
### Feature request ###
**_Most valued users_** list

The _Most valued users_ list would be composed of, let's say people who you don't want to log off often. Who? Backers who have donated more than a million dollars, heads of state, influential professors, etc., and **me**, your first test guinea pig.

Every effort shall be made never to log them off, except as a once a year security exercise with a notice on their screens (thus turning getting logged off into a happy experience (staff cares about my security.))

They will have their own private memcache pool or whatever behind the scenes to assure them that OSM is running smoothly and their donations were put to good use. Of course they would never hear the word memcache. All they know is the site is up.

If the program is a success it would be quietly expanded behind the scenes to eventually encompass all users.

**Comment by jidanni** [comment]
Got logged off again today.

**Comment by jidanni** [comment]
OK, I should just get used to it.

**Comment by swaree** [comment]
Hi, this is happening to me too, I'm being logged off every 3 days or so (even less) as @jidanni reported. Prior to finding these issues   on GitHub (openstreetmap/trac-tickets#5028 this one too) I tried to add an exception to OpenStreetMap for the third-party cookie blocker to no avail; however, this doesn't seem to be the issue. I'd appreciate an update, thanks.

**Comment by tomhughes** [comment]
The update is that we do not guarantee sessions to persist for any particular length of time.

**Comment by washley** [comment]
There's _no_ guarantee at all? What if I get logged off in the middle of making edits? They just get lost?

**Comment by iandees** [comment]
If that happens, your editor will retain the changes and ask you to login again.

**Comment by gy-mate** [comment]
> The update is that we do not guarantee sessions to persist for any particular length of time.

@tomhughes Are there any issues that block guaranteeing session persistence (at least from the server side) for any particular length of time?

It's quite frustrating that the website logs me out every few days. The wiki works fine, it persists my login for months (even after restarting my browser), as I would expect. (I'm aware that's not Ruby, but PHP.)

**Comment by tomhughes** [comment]
I mean the only issue that blocks it is the lack of any technology to do that?

Sessions are currently held in memcached which has a fixed amount of memory so once it's full something has to be evicted.

That said as repeatedly discussed here that is not obviously what is happening so what it really needs is somebody to spend time to diagnose the problem - then we can discuss how it can be fixed.

Short version - there isn't some magic button labelled "annoy everybody by killing sessions" that I can just turn off.

**Comment by pablobm** [comment]
For a contrarian data point: I recently opened the website on my phone after weeks since last time, and it had not logged me out.

**Comment by jidanni** [comment]
Just curious. I've only tried username + password + clicking the Remember Me box.

I notice with these,
<img width="386" height="95" alt="Image" src="https://github.com/user-attachments/assets/ca6b04aa-e3fc-49bb-ac5a-bd41f78439d5" />
there is no Remember Me box. Perhaps if I just choose logging in via one of those,
I won't get logged out so often?

**Comment by tbertels** [comment]
> Just curious. I've only tried username + password + clicking the Remember Me box.
> 
> I notice with these, <img alt="Image" width="386" height="95" src="https://private-user-images.githubusercontent.com/1230959/580382580-ca6b04aa-e3fc-49bb-ac5a-bd41f78439d5.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Nzg3NDY0NzksIm5iZiI6MTc3ODc0NjE3OSwicGF0aCI6Ii8xMjMwOTU5LzU4MDM4MjU4MC1jYTZiMDRhYS1lM2ZjLTQ5YmItYWM1YS1iZDQxZjc4NDM5ZDUuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDUxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA1MTRUMDgwOTM5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGUzN2VmYmM2ZDExOTk4YzIxMzQzNzRlNDU0ZDNmZjI1NmYxZmMwZGE1ZjhhMzhkNGY1OWJkNzhlMDlkMTBhZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmcmVzcG9uc2UtY29udGVudC10eXBlPWltYWdlJTJGanBlZyJ9.NCAd0nvx5ksVlKrWKAoGpGwK04JFzWeJFewahgBgBDE"> there is no Remember Me box. Perhaps if I just choose logging in via one of those, I won't get logged out so often?

No luck there either, I still got disconnected.

**Comment by FormularSumo** [comment]
This is an issue I come across a lot. I get logged out of the OSM website quite frequently, often after only a few days, and this happens on many different devices and browsers (where I'm not logged out of other websites).

I also use other 3rd party apps which log into my OSM account, such as CoMaps or StreetComplete, which as far as I can tell never log me out unless I specifically clear data or reinstall the apps. It seems like the OSM website/ID editor isn't properly saving the user's login session locally to their device, in the same way that these apps are doing.

**Comment by washley** [comment]
Oddly, the last couple of weeks have been great for me, I haven't been logged out once. Before that it was daily 🤷‍♂️
